common/src/main/java/io/pravega/common/concurrent/FutureHelpers.java

```diff
@@ -263,6 +263,33 @@
     }
 
     /**
+     * Same as CompletableFuture.exceptionally(), except that it allows returning a CompletableFuture instead of a single value.
+     *
+     * @param future  The original CompletableFuture to attach an Exception Listener.
+     * @param handler A Function that consumes a Throwable and returns a CompletableFuture of the same type as the original one.
+     *                This Function will be invoked if the original Future completed exceptionally.
+     * @param <T>     Type of the value of the original Future.
+     * @return A new CompletableFuture that will be completed either with the result of future (if it completed normally),
+     * or with the result of handler when applied to the exception of future, should future complete exceptionally.
+     */
+    public static <T> CompletableFuture<T> exceptionallyCompose(CompletableFuture<T> future, Function<Throwable, CompletableFuture<? extends T>> handler) {
+        CompletableFuture<T> result = new CompletableFuture<>();
+        future.thenAccept(result::complete);
+        future.exceptionally(ex -> {
+            try {
+                FutureHelpers.completeAfter(() -> handler.apply(ex), result);
+            } catch (Throwable ex2) {
```

Comment:
really minor, but If you are going to make further changes to this PR, could you please replace `ex2` with something like `innerException`, `innerEx`, or simply something else that isn't a number?

Reply:
fixed